### PR TITLE
refactor(native-filters): update dataMask and ExtraFormData schema

### PR DESCRIPTION
**Pull request by simcha90**
### SUMMARY
This PR implements native filter and cross filter API simplification efforts introduced in https://github.com/apache-superset/superset-ui/pull/1040 and https://github.com/apache-superset/superset-ui/pull/1053 . The following changes are done:

1. `setDataMask` hook now has new structure:
```
{
   filterState: { value: any, ...otherProps: any }
   ownState: { ...otherProps: any }
   extraFormData: ExtraFormData
}
```
1. Nested `append_form_data` and `override_form_data` fields in `ExtraFormData` type are moved up one level into the main filter object
1. `currentValue` removed from formData now it passed in props.filterState.value
1. In SuperChart injected `ownState` and `filterState`
1. Fixed bug when remove cross filter from dashboard it's wasn't from dashboard metadata
1. Fixed bug that `filter_select` was attached to filterBoxes
1. Fixed bug that when user just added cross filter without setting, it's scoping filter not appeared in filter badge
1. migrate old filter set metadata to new format
1. upgrade Python code for legacy charts that depend on `viz.py`

Affected areas:
1. Native filters
2. Cross filters
3. Own state of charts
4. Filter sets

### TEST PLAN
All frontend + backend unit tests have been updated, and a test for the migration has been added to ensure that both up and down migrations produce expected results.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [x] Runtime estimates and downtime expectations provided: only affects dashboards with filter sets. Will only affect environments running the `DASHBOARD_NATIVE_FILTERS_SET` feature flag
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by amitmiran137** [comment]
LGTM, Let's see what CI reveals

**Comment by villebro** [comment]
FYI: marking this as WIP until all `superset-ui` dependencies have been resolved

File: superset-frontend/src/dashboard/actions/dashboardState.js [hunk 1]
```diff
@@ -199,6 +201,28 @@ export function saveDashboardRequest(data, id, saveType) {
...
+                )
+              ) {
+                return { ...prev, [next.id]: next };
+              }
+              return prev;
+            },
+            {},
+          );
+          dispatch(setChartConfiguration(chartConfiguration));
+        }
```
Comment (villebro) [inline superset-frontend/src/dashboard/actions/dashboardState.js:225]: I assume this isn't strictly related to this PR? As this is a pretty heavy PR, could we break this out into a separate PR? (trying to avoid unnecessary bycatches)

File: superset-frontend/src/dataMask/types.ts [hunk 2]
```diff
@@ -16,29 +16,14 @@
...
 
-export type MaskWithId = Mask & { id: string };
-export type DataMaskUnitWithId = { [filterId: string]: MaskWithId };
-export type DataMaskStateWithId = {
-  [DataMaskType.NativeFilters]: DataMaskUnitWithId;
-  [DataMaskType.CrossFilters]: DataMaskUnitWithId;
-  [DataMaskType.OwnFilters]: DataMaskUnitWithId;
-};
+export type DataMaskWithId = DataMask & { id: string };
+export type DataMaskStateWithId = { [filterId: string]: DataMaskWithId };
```
Comment (villebro) [inline superset-frontend/src/dataMask/types.ts:29]: Much better!

**Review by villebro — COMMENTED** [review]
A few comments. During testing I also noticed that filter sets don't seem to work: 
filter-set-refactor

File: superset/utils/core.py [hunk 3]
```diff
@@ -1061,39 +1066,36 @@ def merge_extra_form_data(form_data: Dict[str, Any]) -> None:
...
-        "time_range": "__time_range",
-    }
-    allowed_extra_overrides: Dict[str, Optional[str]] = {
-        "time_grain_sqla": "__time_grain",
-        "druid_time_origin": "__time_origin",
-        "time_range_endpoints": None,
-    }
-
-    applied_time_extras = form_data.get("applied_time_extras", {})
-    form_data["applied_time_extras"] = applied_time_extras
```
Comment (villebro) [inline superset/utils/core.py:1076]: `applied_time_extras` are no longer needed here

**Review by dpgaspar — COMMENTED** [review]
Looks good, left a couple of comments

**Comment by kgabryje** [comment]
LGTM! It seems that there's still the old structure used in `findExistingFilterSet.test.ts`

File: superset/migrations/versions/fc3a3a8ff221_migrate_filter_sets_to_new_format.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    dashboards = (
+        session.query(Dashboard)
+        .filter(Dashboard.json_metadata.like('%"filter_sets_configuration"%'))
+        .all()
+    )
+    changed_filter_sets, changed_filters = 0, 0
+    for dashboard in dashboards:
```
Comment (zhaoyongjie) [inline superset/migrations/versions/fc3a3a8ff221_migrate_filter_sets_to_new_format.py:182]:
```suggestion
    dashboards = (
        session.query(Dashboard)
        .filter(Dashboard.json_metadata.like('%"filter_sets_configuration"%'))
    )
    changed_filter_sets, changed_filters = 0, 0
    for dashboard in dashboards.yield_per(100):
```

File: superset/migrations/versions/fc3a3a8ff221_migrate_filter_sets_to_new_format.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
```
Comment (zhaoyongjie) [inline superset/migrations/versions/fc3a3a8ff221_migrate_filter_sets_to_new_format.py:218]:
```suggestion
    dashboards = (
        session.query(Dashboard)
        .filter(Dashboard.json_metadata.like('%"filter_sets_configuration"%'))
    )
    changed_filter_sets, changed_filters = 0, 0
    for dashboard in dashboards.yield_per(100):
```

**Review by zhaoyongjie — APPROVED** [review]
LGTM! Added some suggestions.

File: superset/migrations/versions/fc3a3a8ff221_migrate_filter_sets_to_new_format.py (hunk 4 above)
Comment (villebro) [reply superset/migrations/versions/fc3a3a8ff221_migrate_filter_sets_to_new_format.py:182]: Thanks - this is a great pattern! As this is behind a FF, I don't expect we'll have lots of these dashboards. Also in the interest of CI time and unblocking other PRs that are dependent on this I think I'll just merge now, but I can open up a follow up PR to implement this.
